### ISSUE TITLE
Stale `inlayHints` in worksheets

### ISSUE THREAD

**Issue by ckipp01** [scalameta/metals#7050]
### Describe the bug

If I have a piece of code like this:

```scala
val x = "hello"
```
I will correctly get the inlay hints I'd expect. However If I change `hello` to `3` like this:

```scala
val x = 3
```

but don't save the buffer yet I get sent inlay hints again that look like this:

```
[Trace - 09:08:10 PM] Sending response 'textDocument/inlayHint - (53)'. Processing request took 24ms
Result: [
  {
    "position": {
      "line": 1,
      "character": 5
    },
    "label": [
      {
        "value": ": "
      },
      {
        "value": "Int"
      }
    ],
    "kind": 1,
    "data": {
      "uri": "file:///Users/ckipp/Documents/scala-workspace/inlay-test/src/main/scala/example/example.worksheet.sc",
      "labelParts": [
        {
          "dataType": "string",
          "string": ""
        },
        {
          "dataType": "string",
          "string": "scala/Int#"
        }
      ]
    }
  },
  {
    "position": {
      "line": 1,
      "character": 9
    },
    "label": " // : String \u003d \"hello\"",
    "tooltip": "x: String \u003d \"hello\""
  }
]
```

Notice that it still includes the `hello` hints at the bottom causing the hints to look not right in your editor.

### Expected behavior

When these hints are sent it should only send them for what is currently in the buffer, not text that has been removed.

### Operating system

macOS

### Editor/Extension

Nvim (nvim-metals)

### Version of Metals

latest snapshot

**Comment by kasiaMarek** [comment]
Would you want the worksheet to evaluate on `textDocument/inlayHint` if the content is stale? Or should it send empty `inlay` hints until new evaluation is triggered (on save)?

**Comment by kasiaMarek** [comment]
```Scala
val x = "aaa" //: String = "aaa"
val y = x //: String = "aaa"
```
```Scala
val x = "bbb"
val y = x
```
That would require some sort of dependency graph for defined values and code blocks.

**Comment by tgodzik** [comment]
Might not be worth tackling then, maybe we can add some decoration to the file like `(stale)` or add it to the status?

**Comment by kasiaMarek** [comment]
Here is a summary from a brief discussion about this issue. We see the following possible solution:

1. Reimplement worksheets, so they are REPL style (each expression top to bottom is fed to REPL). This way all inlay hints below modified part would be outdated (and removed until revaluation). This would make sense if [notebook support](https://github.com/scalameta/metals-feature-requests/issues/236) was added to Metals, since the same underlaying mechanism could be used for both.

2. Reevaluate worksheets when there is a longer break between `didChange` events. This would still sometimes show outdated inlay hints but the result would be fixed during longer periods between typing.

**Comment by masonedmison** [comment]
@kasiaMarek Hrm, well, it seems like `2` would be the easier, would you agree? I don't have any alternate solutions in mind, but I'm happy to have a deeper think if you all aren't satisfied with `1` or `2`.

**Comment by kasiaMarek** [comment]
> it seems like 2 would be the easier, would you agree?

Should be way easier. 😄 

> Though 1 seems to be the more "robust" of the two.

Yes and it is a step towards notebook cells, so it has an additional benefit. But doing this may be a pretty big task.

> I don't have any alternate solutions in mind, but I'm happy to have a deeper think if you all aren't satisfied with 1 or 2.

I think either one should address the issue (to some extent at least), but if you have a better idea we're open to that too. 😄   There was also an idea to clear the "lines" that are outdated, but for code that it not REPL style, it would only end up being a horrific heuristic with more issues than benefits so we let that go.

**Comment by tgodzik** [comment]
Not really, but for Scala 3 this would mean probably using something around https://github.com/scala/scala3/blob/main/compiler/src/dotty/tools/repl/ReplDriver.scala or for Scala 2 around https://github.com/scala/scala/blob/4814abf0106d3150050880b564e9db9d5f0beaf2/src/repl/scala/tools/nsc/interpreter/Interface.scala#L76

We would need to have logic that invokes statements withing the REPL for each statement in the worksheet.

**Comment by tgodzik** [comment]
We would also need to add a new method to PresentationCompiler interface that would handle it.

**Comment by masonedmison** [comment]
I see. What do you mean vy "The normal Parser might be more tricky?"

**Comment by masonedmison** [comment]
@tgodzik I see, thanks!

Question regarding the `WorksheetProvider`...

Is it correct that 
1. compilation of worksheets happens using [ScalaLazyComiler](https://github.com/masonedmison/metals/blob/7050/metals/src/main/scala/scala/meta/internal/metals/Compilers.scala#L1294), 
 and that, 
2. the` `WorksheetProvider`'s main (sole?) responsibility to publish `inlayHints`?

**Comment by tgodzik** [comment]
> compilation of worksheets happens using [ScalaLazyCompiler](https://github.com/masonedmison/metals/blob/7050/metals/src/main/scala/scala/meta/internal/metals/Compilers.scala#L1294)

Currently, it's being evaluated using https://github.com/scalameta/mdoc which creates a file, compiles it and then runs it as a whole. There is a lot of instrumentation done there to get as much data as possible. But basically we generate a fully compilable file from the worksheet, run it and then get back all the intermediate values.

The ScalaLazyCompiler is used to get completions, hovers, all the intellisense feature, but currently doesn't run the worksheet.

> the WorksheetProvider's main (sole?) responsibility is to publish inlayHints?

That's the last part, it does need to compile and run the worksheet.

**Comment by masonedmison** [comment]
I see, hrm, if we moved towards using the Repl to get the inlay hints, would you think we would still use MDoc to compile and run the file?
